### ISSUE TITLE
Make ProgrammaticSuppressionInfo public

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#69000]
## Background and Motivation

We added support for `DiagnosticSuppressor` API some time back. This work added a new internal type `ProgrammaticSuppressionInfo` and an internal property `Diagnostic.ProgrammaticSuppressionInfo` to the compiler layer. This issue proposes making these two APIs public, ~~without any changes or additions to the implementation~~ with one minor change to the existing internal API shape: change the type of [ProgrammaticSuppressionInfo.Suppressions](https://github.com/dotnet/roslyn/blob/eee6243a578fa52f88bc34ccbfa33356797fef94/src/Compilers/Core/Portable/Diagnostic/ProgrammaticSuppressionInfo.cs#L15) property from `ImmutableHashSet<(string Id, LocalizableString Justification)>` to `ImmutableArray<Suppression>`. See existing definition of public type `Suppression` [here](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs)

**Motivation:**
1. Testing support: From @sharwell:
      > We need this public API for `DiagnosticSuppressor` unit testing support in the Microsoft.CodeAnalysis.Testing SDK. We tried two other approaches using public APIs but they caused problems for various edge cases. The library now uses the current property via reflection, which is fragile.

2. Tooling support: Tools built on top of Microsoft.CodeAnalysis that require to read/operate on the programmatic suppressions from `DiagnosticSuppressors` would need this public API. 

### Existing internal type [ProgrammaticSuppressionInfo](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Diagnostic/ProgrammaticSuppressionInfo.cs) to be made public:

```diff
namespace Microsoft.CodeAnalysis.Diagnostics
{
+     public sealed class ProgrammaticSuppressionInfo : IEquatable<ProgrammaticSuppressionInfo?>
+     {
+        // NOTE: The below property is currently implemented in Roslyn with type `ImmutableHashSet<(string Id, LocalizableString Justification)>`, which might not be best suited for a public API.
+        //       Also note that this is proposed as an array and not a single Suppression instance because multiple diagnostic suppressors can suppress the same diagnostic. For most real world cases, the below array will have a single element.
+        public ImmutableArray<Suppression> Suppressions { get; }

+        public bool Equals(ProgrammaticSuppressionInfo? other);
+        public overrides bool Equals(object? other);
+        public overrides int GetHashCode();
+     }
}
```

### Existing internal property [Diagnostic.ProgrammaticSuppressionInfo](https://github.com/dotnet/roslyn/blob/86156faa88860dac9b17223623743aea24814291/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs#L474) to be made public:

```diff
namespace Microsoft.CodeAnalysis
{
      public abstract class Diagnostic
      {
+        public virtual ProgrammaticSuppressionInfo? ProgrammaticSuppressionInfo { get; };
      }
}
```

**NOTE:** `Diagnostic` type has internal abstract methods, and hence cannot be sub-typed outside the compiler layer: https://github.com/dotnet/roslyn/blob/86156faa88860dac9b17223623743aea24814291/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs#L448-L461
```

**Comment by mavasani** [comment]
@sharwell @333fred

**Comment by sharwell** [comment]
Also note this API has existed in its current form for a long time. When making it public, it would benefit consumers of the API if we avoid making any changes to any names or return values.

**Comment by mavasani** [comment]
> ```diff
> public ImmutableHashSet<(string Id, LocalizableString Justification)> Suppressions { get; }
> ```

I think this should be possible for most of the existing API shape. The only thing that needs some discussion is the return type of `Suppressions` property: https://github.com/dotnet/roslyn/blob/eee6243a578fa52f88bc34ccbfa33356797fef94/src/Compilers/Core/Portable/Diagnostic/ProgrammaticSuppressionInfo.cs#L15

Currently, it is an immutable hash set of tuple values. I believe we can consider following alternatives:
- Change the property type to be an `ImmutableArray` instead of `ImmutableHashSet`
- Introduce a new public struct type, say `ProgrammaticSuppresion` with 2 fields

Basically:

```csharp
public sealed class ProgrammaticSuppressionInfo : IEquatable<ProgrammaticSuppressionInfo?>
{
    public ImmutableArray<ProgrammaticSuppression> Suppressions { get; }
    ...
}

public readonly struct ProgrammaticSuppression
{
    public string SuppressionId { get; }
    public LocalizableString Justification { get; }
}
```

I'll add this alternate proposal to the original issue description

**Comment by 333fred** [comment]
### API Review

* Can we move `Suppressions` to `SuppressionInfo` instead and not have `ProgrammaticSuppressionInfo` at all?
    * Let's see if we can consolidate into `SuppressionInfo`. Passing a `Compilation` might be a problem?

**Conclusion**: See if we can do that move, and send email to the api review alias to follow up with the results. If it can't be done, we can make sure we like the names proposed and then approve over email.

**Comment by mavasani** [comment]
> Can we move Suppressions to SuppressionInfo instead and not have ProgrammaticSuppressionInfo at all

This seems doable, I'll try to work up a draft PR to confirm before moving this issue back to API review.

**Comment by mavasani** [comment]
@333fred Marking this ready for review with the proposed API addition as per the suggestion in https://github.com/dotnet/roslyn/issues/69000#issuecomment-1664646444.

**Draft PR:** https://github.com/dotnet/roslyn/pull/71207

**Proposed API addition:**

```diff
namespace Microsoft.CodeAnalysis.Diagnostics
{
      public sealed class SuppressionInfo
      {
+          /// <summary>
+          /// If the diagnostic was suppressed by one or more programmatic suppressions by <see cref="DiagnosticSuppressor"/>(s),
+          /// then returns the corresponding set of <see cref="Suppression"/>s.
+          /// Otherwise, returns an empty set.
+          /// </summary>
+          public ImmutableHashSet<Suppression> Suppressions { get; }
      }
}
```

**Existing `SuppressionInfo` definition:** https://github.com/dotnet/roslyn/blob/d41a105acac6a8822206b8b142eab881d88645d4/src/Compilers/Core/Portable/Diagnostic/SuppressionInfo.cs#L5-L29

**Comment by 333fred** [comment]
### API Review

* Should `Suppressions` be an `ImmutableArray` instead?
    * Doesn't seem like there are any blockers, just comment that order isn't guaranteed?
    * Unless we're already using it as a HashSet internally and it's a perf hit to use an array.
    * We'll let this be driven by the implementation; whichever is better for it.
* Change the property name to `ProgrammaticSuppressions`

**Approved with name/return type change**

**Comment by mavasani** [comment]
> Unless we're already using it as a HashSet internally and it's a perf hit to use an array.

We should be able to switch over to an array without a perf hit.

**Comment by mikechristiansenvae** [comment]
@mavasani : I think `ProgrammaticSuppressionInfoWrapper` in roslyn-sdk should be changed to `ImmutableArray` as well.

https://github.com/dotnet/roslyn-sdk/blob/f500e81ba5596809e711a3022bf8e80a00c8371b/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/Lightup/ProgrammaticSuppressionInfoWrapper.cs#L29

Otherwise, this occurs:

```
System.TypeInitializationException: The type initializer for 'Microsoft.CodeAnalysis.Testing.Lightup.ProgrammaticSuppressionInfoWrapper' threw an exce...

System.TypeInitializationException
The type initializer for 'Microsoft.CodeAnalysis.Testing.Lightup.ProgrammaticSuppressionInfoWrapper' threw an exception.
   at Microsoft.CodeAnalysis.Testing.Lightup.ProgrammaticSuppressionInfoWrapper.FromInstance(Object instance) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/Lightup/ProgrammaticSuppressionInfoWrapper.cs:line 32
   at Microsoft.CodeAnalysis.Testing.Extensions.DiagnosticExtensions.ProgrammaticSuppressionInfo(Diagnostic diagnostic) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/Extensions/DiagnosticExtensions.cs:line 38
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.<>c.<FilterDiagnostics>b__103_0(ValueTuple`2 d) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 1687
   at System.Linq.Enumerable.ArrayWhereIterator`1.ToArray(ReadOnlySpan`1 source, Func`2 predicate)
   at System.Linq.Enumerable.ArrayWhereIterator`1.ToArray()
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at System.Collections.Immutable.ImmutableArray.CreateRange[T](IEnumerable`1 items)
   at System.Collections.Immutable.ImmutableArray.ToImmutableArray[TSource](IEnumerable`1 items)
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.FilterDiagnostics(ImmutableArray`1 diagnostics) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 1686
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.GetSortedDiagnosticsAsync(Solution solution, ImmutableArray`1 analyzers, ImmutableArray`1 additionalDiagnostics, CompilerDiagnostics compilerDiagnostics, IVerifier verifier, CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 1144
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.GetSortedDiagnosticsAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, ImmutableArray`1 analyzers, IVerifier verifier, CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 1090
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.VerifyDiagnosticsAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, DiagnosticResult[] expected, IVerifier verifier, CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 449
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.RunImplAsync(CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 203
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.RunAsync(CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 172
   at MyOrg.Analyzers.Tests.MyOrg0006_AllowedUnusedPrivateMembers.RunTestAsync(String source, DiagnosticResult[] expected) in S:\projects\my-repo\source\MyOrg.Analyzers\tests\MyOrg.Analyzers.Tests\MyOrg0006_AllowedUnusedPrivateMembers.cs:line 90
   at MyOrg.Analyzers.Tests.MyOrg0006_AllowedUnusedPrivateMembers.ExpectNoDiagnostic() in S:\projects\my-repo\source\MyOrg.Analyzers\tests\MyOrg.Analyzers.Tests\MyOrg0006_AllowedUnusedPrivateMembers.cs:line 75
   at Xunit.v3.TestRunner`2.<>c__DisplayClass5_0.<<InvokeTest>b__1>d.MoveNext() in /_/src/xunit.v3.core/Runners/TestRunner.cs:line 170
--- End of stack trace from previous location ---
   at Xunit.v3.ExceptionAggregator.RunAsync(Func`1 code) in /_/src/xunit.v3.core/Exceptions/ExceptionAggregator.cs:line 124

System.InvalidOperationException
Property 'System.Collections.Immutable.ImmutableArray`1[Microsoft.CodeAnalysis.Diagnostics.Suppression] Suppressions' produces a value of type 'System.Collections.Immutable.ImmutableArray`1[Microsoft.CodeAnalysis.Diagnostics.Suppression]', which is not assignable to type 'System.Collections.Immutable.ImmutableHashSet`1[System.ValueTuple`2[System.String,Microsoft.CodeAnalysis.LocalizableString]]'
   at Microsoft.CodeAnalysis.Testing.Lightup.LightupHelpers.CreatePropertyAccessor[T,TResult](Type type, String propertyName, TResult defaultValue) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/Lightup/LightupHelpers.cs:line 53
   at Microsoft.CodeAnalysis.Testing.Lightup.ProgrammaticSuppressionInfoWrapper..cctor() in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/Lightup/ProgrammaticSuppressionInfoWrapper.cs:line 21
```

**Comment by CyrusNajmabadi** [comment]
~~@mikechristiansenvae that code was moved into this repo.  It was updated to reference the correct types in https://github.com/dotnet/roslyn/pull/79030, https://github.com/dotnet/roslyn/pull/79024 and https://github.com/dotnet/roslyn/pull/79008~~

**Comment by CyrusNajmabadi** [comment]
Ignore me.  got my wires crossed.
